extract_pov_edits/dataflow_main.py

```diff
@@ -0,0 +1,370 @@
+    new_seq = {}
+    equals = []
+    inserts = []
+    deletes = []
+    change_in_new = []
+    for s in sents:
+      if s in sents_old:
+        new_seq[s] = cur_sents[s]
+        equals.append((s, cur_sents[s]))
+        change_in_new.append(False)
+      else:
+        new_seq[s] = rev_id
+        inserts.append((s, rev_id))
+        change_in_new.append(True)
+    change_in_old = []
+    for s in sents_old:
+      if s not in sents:
+        change_in_old.append(True)
+        deletes.append((s, cur_sents[s]))
+      else:
+        change_in_old.append(False)
+    # Locate context sentences that are unchanged near the inseted/deleted
+    # sentences.
+    context_equals = []
+    for ind, s in enumerate(sents):
+      if (not change_in_new[ind] and
+          self.change_close_by(CONTEXT_RANGE, ind, change_in_new)):
+          context_equals.append(s)
+    for ind, s in enumerate(sents_old):
+      if (not change_in_old[ind] and
+          self.change_close_by(CONTEXT_RANGE, ind, change_in_old)):
+          context_equals.append(s)
+    # Dedeuplicate sentences.
+    context_equals = list(set(context_equals))
+    return context_equals, inserts, deletes, new_seq
+
+
+  def format_clean(self, s):
+    """Clean MediaWiki format."""
+    # Clean titles
+    ret = []
+    for line in s.splitlines():
+      front_cnt = re.search("^=+", line)
+      back_cnt = re.search("^=+", line[::-1])
+      if (front_cnt is None or back_cnt is None
+          or len(front_cnt.group(0)) != len(back_cnt.group(0))):
+        ret.append(line)
+    s = "\n".join(ret)
+    # mwparserfromhell clean.
+    s = mwparserfromhell.parse(s).strip_code()
+    return s
+
+
+  def process(self, element, bucket, ingestFrom):
+    """Ingests the xml dump into json, returns the josn records
+    """
+    nltk.download('punkt')
+    # Decompress the data dump
+    chunk_name = element
+    logging.info('USERLOG: Running ingestion process on %s' % chunk_name)
+    if ingestFrom == 'local':
+       input_stream = chunk_name
+    else:
+       os.system("gsutil -m cp %s %s" % (path.join('gs://', bucket, chunk_name), chunk_name))
+       input_stream = chunk_name
+    # Running ingestion on the xml file
+    last_revision = 'None'
+    last_completed = time.time()
+    cur_sents = {}
+    i = 0
+    for i, content in enumerate(parse_stream(bz2.BZ2File(chunk_name))):
+      self.processed_revisions.inc()
+      # Add the week and year field for sharding
+      if content['text'] is None:
+        content['text'] = ""
+      content['text'] = self.format_clean(content['text'])
+      last_revision = content['rev_id']
+      for line in content['text'].split('\n'):
+        if len(line) <= LINE_THERESHOLD:
+           sents = nltk.sent_tokenize(line)
+        else:
+           beam.pvalue.TaggedOutput('error_log', json.dumps(content['rev_id']))
+      context_equals, inserts, deletes, cur_sents = self.diff(sents, cur_sents, content['rev_id'])
```

Comment:
This happens even if you are the threshold, in which case you are using the sents from last loop iteration.